### ISSUE TITLE
Cannot Read property 'lenght' of Undefined

### ISSUE THREAD

**Issue by vonboth** [fullcalendar/fullcalendar#3598]
Updated to FC 3.3.0 just now.
Unfortunately FC breaks working with event sources

I work with two sources:

```
eventSources: [
        {
          events: function(start, end, timezone, callback) {
            _getAdates(start, end, timezone, callback);
          }
        },
        {
          events: function(start, end, timezone, callback) {
            _getAdaysPeriods(start, end, timezone, callback);
          }
        }
      ],
```

Both are $.ajax calls.
In the success function I call: callback(events)
The Ajax call is successfull as I can see from the broswers timeline.

However, fullcalendar breaks with this stack trace:

```
Uncaught TypeError: Cannot read property 'length' of undefined
    at members.constructor.renderEvents (fullcalendar.js:4545)
    at members.constructor.renderEvents (fullcalendar.js:14164)
    at fullcalendar.js:9544
    at Object.promise.then (fullcalendar.js:1969)
    at members.constructor.executeEventsRender (fullcalendar.js:9543)
    at fullcalendar.js:9504
    at runFunc (fullcalendar.js:2033)
    at fullcalendar.js:2050
    at new Promise (fullcalendar.js:1908)
    at TaskQueue.addTask (fullcalendar.js:2028)
    at members.constructor.requestEventsRender (fullcalendar.js:9503)
    at members.constructor.handleEvents (fullcalendar.js:9486)
    at members.constructor.setEvents (fullcalendar.js:9453)
    at fullcalendar.js:9429
    at Object.Promise.promise.then (fullcalendar.js:1929)
    at members.constructor.bindEvents (fullcalendar.js:9427)
    at fullcalendar.js:8987
    at Object.Promise.promise.then (fullcalendar.js:1929)
    at members.constructor.handleDate (fullcalendar.js:8985)
    at members.constructor.setDate (fullcalendar.js:8963)
    at renderView (fullcalendar.js:11652)
    at Calendar_constructor.render (fullcalendar.js:11523)
    at HTMLDivElement.<anonymous> (fullcalendar.js:43)
    at Function.each (jquery.min.js:2)
    at r.fn.init.each (jquery.min.js:2)
    at r.fn.init.$.fn.fullCalendar (fullcalendar.js:35)
    at HTMLDivElement.windowResize (CalendarViewModel.js:152)
    at Calendar_constructor.publiclyTrigger (fullcalendar.js:12049)
    at members.constructor.publiclyTrigger (fullcalendar.js:8759)
    at windowResize (fullcalendar.js:11781)
    at later (fullcalendar.js:1068)
```

**Comment by arshaw** [comment]
Would you be able to post a [runnable, stripped-down demonstration of the bug](http://fullcalendar.io/wiki/Reduced-Test-Cases/)? Would really appreciate it because the time saved reproducing will be time spent fixing.

**Comment by micahbrich** [comment]
I caught this bug, too. I can't strip it down without making a huge mess, but I'm also calling $.ajax, so maybe there's some commonality.

**Comment by jefbarn** [comment]
Also just hitting this now with async event fetching.  
Verified that it's a break in 3.3.0.
Reverted to 3.2 works fine.
@arshaw can you reopen?

**Comment by espen** [comment]
Can anyone reproduce this in a [reduced test case](https://fullcalendar.io/wiki/Reduced-Test-Cases/)?

**Comment by jefbarn** [comment]
I tried doing a jsfiddle but wasn't reproducing, using similar code in my app does. I'll have to dig into it some more, but my guess is a race condition in the event loading code.

**Comment by vonboth** [comment]
Hi Adam,
I tried to create a fiddle, but ran into another problem:
https://jsfiddle.net/vonboth/01ur7a9a/

Trace:

```
Uncaught TypeError: Cannot read property '_calendar' of undefined
    at D (moment.min.js:6)
    at e (VM237 jquery.min.js:4)
    at Ab (VM237 jquery.min.js:4)
    at Ab (VM237 jquery.min.js:4)
    at Ab (VM237 jquery.min.js:4)
    at Function.r.param (VM237 jquery.min.js:4)
    at Function.ajax (VM237 jquery.min.js:4)
    at _t.events (VM240:66)
    at f (fullcalendar.min.js:6)
    at c (fullcalendar.min.js:6)
    at u (fullcalendar.min.js:6)
    at a (fullcalendar.min.js:6)
    at o (fullcalendar.min.js:6)
    at _t.n [as requestEvents] (fullcalendar.min.js:6)
    at function.e.constructor.requestEvents (fullcalendar.min.js:9)
    at function.e.constructor.bindEvents (fullcalendar.min.js:9)
    at fullcalendar.min.js:9
    at Object.Et.i.then (fullcalendar.min.js:6)
    at function.e.constructor.handleDate (fullcalendar.min.js:9)
    at function.e.constructor.setDate (fullcalendar.min.js:9)
    at u (fullcalendar.min.js:6)
    at o (fullcalendar.min.js:6)
    at _t.s [as render] (fullcalendar.min.js:6)
    at HTMLDivElement.<anonymous> (fullcalendar.min.js:6)
    at Function.each (VM237 jquery.min.js:2)
    at r.fn.init.each (VM237 jquery.min.js:2)
    at r.fn.init.t.fn.fullCalendar (fullcalendar.min.js:6)
    at HTMLDocument.<anonymous> (VM240:63)
    at j (VM237 jquery.min.js:2)
    at k (VM237 jquery.min.js:2)
```
Which now seems to break in moment
It even doesn't start the request at all.

I agree with jefbarn, that the problem is some race of resolving Promises.

What I do in my code ist the following:

```
events: function() {

var eventsLoader = $.ajax({
        url: 'some/url',
        method: 'get',
        type: 'JSON',
        data: {start: start, end: end}
 })

var colorLoader = $.ajax .......

    $.when(colorLoader, eventsLoader)
      .then(function (resolvedColors, resolvedEvents) {
        var coloredEvents = self.setEventBackgroundColorAndLegend(
          resolvedEvents[0].data.events,
          resolvedColors[0].data.colors
        );
        callback(coloredEvents); //hand over colored events to Fullcalendar

      })
      .fail(function (error) {
        console.error(error);
      });
}
```
This fails in FC before all requests are resolved, but they are resolved properly as I can see from my dev-tools

**Comment by jefbarn** [comment]
@vonboth I modified your jsfiddle to use fetch, but the problem doesn't reproduce...
https://jsfiddle.net/01ur7a9a/4/

**Comment by jefbarn** [comment]
Seems that 'gotoDate' is the cause...
https://jsfiddle.net/jefbarn/01ur7a9a/6/

**Comment by arshaw** [comment]
thanks @jefbarn . I'll try to fix the your reported/recreated bug, which involves `gotoDate`.

If anyone else has a fiddle of a different bug, please post.

**Comment by arshaw** [comment]
@jefbarn and @vonboth you seem to have the same bug (`._calendar` is undefined). It's actually with jQuery trying to serialize the moment object for use in the ajax param list. I've stripped away FullCalendar and it's still happening:

https://jsfiddle.net/57pt1hk3/1/

Don't pass a raw moment object to jquery's `data`. use `.format()` on the moment first.

**Comment by arshaw** [comment]
err, that was @vonboth 's issue and recreation i was addressing, not @jefbarn . I'm still going to look into the `gotoDate` problem.

**Comment by vonboth** [comment]
@arshaw + @jefbarn 
I simply started the Chrome debugger and set the breakpoint here:
```
at later (fullcalendar.js:1068)
```
and steped through the code.

FC tries to call windowResize at this step and calls the 'windowResize' callback function from the options.
I've set this to:
```
windowResize: function (view) {
        $('#calendar').fullCalendar('render');
      },
```
If I comment this out (or leave empty), the bug doesn't appear any longer and FC works as expected.
Now this makes sense to me, as I want FC to render my calendar right away while the Promise hasn't resolved and of course the length of the events array is empty.

Hope this might be another starting point for you Adam besides the issue with 
`gotoDate`

Thanks for your help and your stunning work!
